batchlib/mongo/result_writer.py

```diff
@@ -0,0 +1,137 @@
+import glob
+import os
+import urllib.parse
+from datetime import datetime
+
+import git
+import h5py
+from pymongo import MongoClient
+
+from batchlib.base import BatchJobOnContainer
+from batchlib.util import get_logger
+from batchlib.util.io import read_table
+
+ASSAY_ANALYSIS_RESULTS = 'immuno-assay-analysis-results'
+
+logger = get_logger('Workflow.BatchJob.DbResultWriter')
+
+
+def _get_table_names(f):
+    table_names = set()
+
+    def _visitor(name):
+        if name.endswith('cells'):
+            table_names.add(name[:-6])
+
+    f['tables'].visit(_visitor)
+    return list(table_names)
+
+
+def _table_object(column_names, table):
+    # document attributes cannot contain '.'
+    column_names = [cn.replace('.', '_') for cn in column_names]
+    return [
+        dict(zip(column_names, table_row)) for table_row in table
+    ]
+
+
+def _get_result_tables(in_file):
+    with h5py.File(in_file, 'r') as f:
+        result_tables = {}
+        for table_name in _get_table_names(f):
+            column_names, table = read_table(f, table_name)
+            result_tables[table_name] = _table_object(column_names, table)
+
+        return result_tables
+
+
+def _get_log_path(work_dir):
+    logs = list(glob.glob(os.path.join(work_dir, '*.log')))
+    assert len(logs) == 1
+    return logs[0]
+
+
+def _parse_workflow_duration(work_dir):
+    """
+    Reads workflow duration by parsing the first and the last log event in the log file and taking the time difference
+    """
+    log_path = _get_log_path(work_dir)
+    with open(log_path, 'r') as fh:
+        lines = list(fh)
+        for first_log in lines:
+            if 'INFO' in first_log:
+                break
+
+        for last_log in reversed(lines):
+            if 'INFO' in last_log:
+                break
+
+        start = datetime.strptime(first_log.split('[')[0].strip(), '%Y-%m-%d %H:%M:%S')
+        end = datetime.strptime(last_log.split('[')[0].strip(), '%Y-%m-%d %H:%M:%S')
+
+        delta = end - start
+        return delta.seconds
+
+
+def _parse_workflow_name(work_dir):
+    """
+    Parses workflow name from the log file
+    """
+    log_path = _get_log_path(work_dir)
+    logfile = os.path.split(log_path)[1]
+    workflow_name = os.path.splitext(logfile)[0]
+    return workflow_name
+
+
+def _get_git_sha():
+    try:
+        repo = git.Repo(search_parent_directories=True)
+        return repo.head.object.hexsha
+    except Exception as e:
+        logger.warning(f'Cannot get batchlib_version: {e}')
+        return None
+
+
+class DbResultWriter(BatchJobOnContainer):
```

Comment:
You should also over-write `validate_outputs`and `check_outputs`, so that this can be run without `force_recompute`.
Just check whether you have the unique identifier (plate_name + commit id, I think?) for this entry. And `return True if db is None`.

Reply:
awesome! Thanks for the explanation. Now the `force_recompute` it's handled correctly:
we check if the (commit_id, workflow_name, plate_name) id exist in the db (`check_output`) and skip computation if it does and `force_recompute=False`. Otherwise we replace the result object in the DB with the new one. `validate_output` just returns `True` in our case.
I've tested all the different use-cases and it behaves as expected.
:white_check_mark: